0018/README.md

```diff
@@ -0,0 +1,229 @@
+- **Command invocation agent (agent)**: any service that receives endpoint command data from other service(s).
+
+
+## Design
+
+### Command invocation request
+
+*Command invocation request* is a [targeted message](/0003/README.md#targeted-messaging) published by command invocation caller that wants to invoke command on specific endpoint.
+
+The caller MUST send command invocation request messages using the following NATS subject:
+```
+kaa.v1.service.{provider-service-instance-name}.cip.command-request
+```
+
+The caller MUST include NATS `replyTo` field to handle the [command invocation result response](#command-invocation-result).
+It is RECOMMENDED to follow the subject format described in [3/ISM session affinity section](/0003/README.md#session-affinity):
+```
+kaa.v1.replica.{consumer-service-replica-id}.cip.command-result
+```
+
+For more information, see [3/Messaging IPC](/0003/README.md).
+
+
+*Command invocation request* message payload MUST be an [Avro-encoded](https://avro.apache.org/) object with the following schema ([0018-command-invocation-request.avsc](./0018-command-invocation-request.avsc)):
+
+```json
+{
+    "namespace":"org.kaaproject.ipc.cip.gen.v1",
+    "type":"record",
+    "name":"CommandInvocationRequest",
+    "doc":"Command invocation request message.",
+    "fields":[
+        {
+            "name":"correlationId",
+            "type":"string",
+            "doc":"Message ID that is used for message processing tracking across services."
+        },
+        {
+            "name":"timestamp",
+            "type":"long",
+            "doc":"Message creation UNIX timestamp in milliseconds."
+        },
+        {
+            "name":"timeout",
+            "type":"long",
+            "default":0,
+            "doc":"Amount of milliseconds (since the timestamp) until the message expires. Value of 0 is reserved to indicate no expiration."
+        },
+        {
+            "name":"endpointId",
+            "type":"string",
+            "doc":"Identifier of the endpoint the command is invoked against."
+        },
+        {
+            "name":"commandType",
+            "type":"string",
+            "doc":"Command type."
+        },
+        {
+            "name":"commandId",
+            "type":"int",
+            "doc":"Unique command ID that is used for command correlation. (endpointId, commandType, commandId) tuple uniquely identifies a command instance."
+        },
+        {
+            "name":"payload",
+            "type":[
+              "null",
+              "bytes"
+            ],
+            "default":null,
+            "doc":"The command payload to be interpreted by endpoint."
+        }
+    ]
+}
+```
+
+Example:
+
+```json
+{  
+   "correlationId":"07d78e95-2c4d-4899-957c-b9e5a3701fbb",
+   "timestamp":1514372799674,
+   "timeout":0,
+   "endpointId":"b197e391-1d13-403b-83f5-87bdd44888cf",
+   "commandType":"measurement",
+   "commandId":284,
+   "payload":{  
```

Comment:
This does not match Avro schema above. `payload` field must be bytes, not an arbitrary value.

Reply:
Correct example
```json
"payload": { "bytes": "{\"temperature\":true,\"humidity\":false}" }
```